File: Kotlin/ch02/src/main/kotlin/ch02/LinkedListStack.kt
package ch02

/**
 * 연결 리스트 기반 스택 (LinkedListStack)
 * Rust/ch02/src/linked_list_stack.rs → Kotlin 이디엄 포팅
 * 책의 원본 C 코드는 char* 데이터를 저장하므로 String을 저장한다.
 */
class LinkedListStack {

    private val items = ArrayDeque<String>()

    fun push(data: String) = items.addLast(data)

    fun pop(): String? = items.removeLastOrNull()

    fun top(): String? = items.lastOrNull()

    fun size(): Int = items.size

    fun isEmpty(): Boolean = items.isEmpty()
}

fun main() {
    val stack = LinkedListStack()
    stack.push("abc")
    stack.push("def")
    stack.push("efg")

    while (!stack.isEmpty()) println("Popped: ${stack.pop()}")
}

File: Rust/ch02/src/bin/linked_list_stack.rs
// 연결 리스트 스택 테스트
// Clang/02/LinkedListStack/Test_LinkedListStack.c 포팅.
use ch02::linked_list_stack::LinkedListStack;

fn main() {
    let mut stack = LinkedListStack::new();

    stack.push("abc".to_string());
    stack.push("def".to_string());
    stack.push("efg".to_string());
    stack.push("hij".to_string());

    let count = stack.size();
    println!("Size: {}, Top: {}\n", count, stack.top().unwrap());

    for _ in 0..count {
        if stack.is_empty() {
            break;
        }

        let popped = stack.pop().unwrap();
        print!("Popped: {}, ", popped);

        if !stack.is_empty() {
            println!("Current Top: {}", stack.top().unwrap());
        } else {
            println!("Stack Is Empty.");
        }
    }
}
